server/artemis/remote.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+let throttleTime: number | null = null;
+export const limiter = new Limiter(15, 10);
+export const PM = new ProcessManager.QueryProcessManager<string, Record<string, number> | null>(module, async text => {
+	if (isCommon(text) || !limiter.shouldRequest()) return null;
+	if (throttleTime && (Date.now() - throttleTime < 10000)) {
+		return null;
+	}
+	if (throttleTime) throttleTime = null;
+
+	const requestData: PerspectiveRequest = {
+		// todo - support 'es', 'it', 'pt', 'fr' - use user.language? room.settings.language...?
+		languages: ['en'],
+		requestedAttributes: ATTRIBUTES,
+		comment: {text},
+	};
+	try {
+		const raw = await Net(`https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze`).post({
+			query: {
+				key: Config.perspectiveKey,
+			},
+			body: JSON.stringify(requestData),
+			headers: {
+				'Content-Type': "application/json",
+			},
+			timeout: 10 * 1000, // 10s
+		});
+		if (!raw) return null;
+		const data = JSON.parse(raw);
+		if (data.error) throw new Error(data.message);
+		const result: {[k: string]: number} = {};
+		for (const k in data.attributeScores) {
+			const score = data.attributeScores[k];
+			result[k] = score.summaryScore.value;
+		}
+		return result;
+	} catch (e: any) {
+		throttleTime = Date.now();
+		if (e.message.startsWith('Request timeout')) {
+			// just ignore this. error on their end not ours.
+			// todo maybe stop sending requests for a bit?
+			return null;
+		}
+		Monitor.crashlog(e, 'A Perspective API request', {request: JSON.stringify(requestData)});
+		return null;
+	}
+}, PM_TIMEOUT);
+
+
+// main module check necessary since this gets required in other non-parent processes sometimes
+// when that happens we do not want to take over or set up or anything
+if (!PM.isParentProcess && require.main === module) {
+	// This is a child process!
+	global.Config = Config;
+	global.Monitor = {
+		crashlog(error: Error, source = 'A remote Artemis child process', details: AnyObject | null = null) {
+			const repr = JSON.stringify([error.name, error.message, source, details]);
+			process.send!(`THROW\n@!!@${repr}\n${error.stack}`);
+		},
+		slow(text: string) {
+			process.send!(`CALLBACK\nSLOW\n${text}`);
+		},
+	};
+	global.toID = toID;
+	process.on('uncaughtException', err => {
+		if (Config.crashguard) {
+			Monitor.crashlog(err, 'A remote Artemis child process');
+		}
+	});
+	// eslint-disable-next-line no-eval
+	Repl.start(`abusemonitor-remote-${process.pid}`, cmd => eval(cmd));
+} else if (PM.isParentProcess) {
+	PM.spawn(Config.remoteartemisprocesses || 1);
+}
+
+export class RemoteClassifier {
+	static readonly PM = PM;
+	static readonly ATTRIBUTES = ATTRIBUTES;
+	classify(text: string) {
+		if (!Config.perspectiveKey) return Promise.resolve(null);
+		return PM.query(text);
+	}
+	destroy() {
+		return PM.destroy();
+	}
+	respawn() {
+		return PM.respawn();
+	}
+	spawn(number: number) {
+		PM.spawn(number);
+	}
+	getActiveProcesses() {
+		return PM.processes.length;
+	}
+}
+
```

Comment:
I assume this file is just copy/pasted from the preexisting Artemis code? Are there changes I should be aware of?

Reply:
No. It's all the same.